runtime/tr.source/trj9/optimizer/J9TransformUtil.cpp

```diff
@@ -214,93 +268,117 @@ static bool verifyFieldAccess(void *curStruct, TR::SymbolReference *field, TR::C
+      if (addressChildNode == baseNode)
+         {
+         // baseStruct is the value of baseNode, dereference is not needed
+         curStruct = (uintptrj_t)baseStruct;
+         // baseStruct/baseNode are deemed verifiable by the caller         }
+         }
+      else
+         {
+         TR::SymbolReference *addressChildSymRef = addressChildNode->getSymbolReference();
+         // Get the address of struct containing current field and dereference it
+         void* addressChildAddress = dereferenceStructPointerChain(baseStruct, baseNode, addressChildNode, comp);
+
+         if (addressChildAddress == NULL)
+            {
+            return NULL;
+            }
+         // Since we're going to dereference addressChildAddress, addressChild must be a java reference or a native struct
+         else if (addressChildSymRef->getSymbol()->isCollectedReference())
+            {
+            curStruct = comp->fej9()->getReferenceFieldAtAddress((uintptrj_t)addressChildAddress);
+            }
+         else // Native struct
+            {
+            TR_ASSERT(isFinalFieldPointingAtNativeStruct(addressChildSymRef, comp), "dereferenceStructPointerChain should be dealing with reference fields");
+            curStruct = *(uintptrj_t*)addressChildAddress;
+            }
+         }
+
+      // Get the field address of curNode
       if (curStruct)
          {
-         if (verifyFieldAccess(curStruct, curNode->getSymbolReference(), comp))
+         if (verifyFieldAccess((void*)curStruct, symRef, comp))
             {
-            if (isJavaObjectReferenceFieldOfJavaObject(curNode->getSymbolReference(), comp))
+            uintptrj_t fieldAddress = 0;
+            // The offset of a java field is in its symRef
+            if (isJavaField(symRef, comp))
                {
-               uintptrj_t fieldAddress = (uintptrj_t)(curStruct + curNode->getSymbolReference()->getOffset());
-               TR_J9VM *vm = (TR_J9VM*)(comp->fe());
-               return (void*)vm->getReferenceFieldAtAddress(fieldAddress);
+               fieldAddress = curStruct + symRef->getOffset();
+               }
+            else if (comp->getSymRefTab()->isImmutableArrayShadow(symRef))
+               {
+               TR::Node* offsetNode = curNode->getFirstChild()->getSecondChild();
+               if (!offsetNode->getOpCode().isLoadConst())
+                  return NULL;
+
+               int64_t offset = 0;
+               if (offsetNode->getDataType() == TR::Int64)
+                  offset = offsetNode->getUnsignedLongInt();
+               else
+                  offset = offsetNode->getUnsignedInt();
+
+               // Calculate index from offset
+               int32_t index = TR::TransformUtil::calculateIndexFromOffsetInContiguousArray(comp, offset, curNode->getDataType());
+               int32_t arrayLength = comp->fej9()->getArrayLengthInElements(curStruct);
+
+               // Check array bound
+               if (index < 0 || index >= arrayLength)
```

Comment:
If the type of `curNode` isn't the same size as the element type of the array, then `index` and `arrayLength` will be in different units. For example, if we have a `byte[]` of length 8, and `curNode` is `sloadi` from offset (8 + array header size), then `arrayLength=8` and `index=4`, which appears to be in bounds.

Because of `verifyFieldAccess()`, this can only happen when both are primitive types, and therefore `curNode` is the outermost load in the chain. It would be fine to load garbage in this case, but it's possible that the array is near the end of a page, and the next page may not be accessible

Reply:
A change has been made to compare `offset` instead of `index`.